File: services/server/src/status/controller/status.controller.ts
import {
  Controller,
  Get,
  Post,
  Delete,
  Param,
  Body,
  UseGuards,
  HttpCode,
  HttpStatus,
} from "@nestjs/common";
import { ApiTags, ApiOperation, ApiBearerAuth } from "@nestjs/swagger";
import { JwtAuthGuard } from "@/auth/controller/jwt-auth.guard";
import { CurrentUser } from "@/auth/controller/current-user.decorator";
import { StatusService } from "@/status/service/status.service";

@ApiTags("status")
@Controller("status")
@UseGuards(JwtAuthGuard)
@ApiBearerAuth()
export class StatusController {
  constructor(private readonly statusService: StatusService) {}

  @Get()
  @ApiOperation({ summary: "List statuses" })
  async getStatuses(@CurrentUser() user: { id: string }) {
    const statuses = await this.statusService.getStatuses(user.id);
    return { statuses };
  }

  @Post()
  @HttpCode(HttpStatus.CREATED)
  @ApiOperation({ summary: "Create status" })
  async createStatus(
    @CurrentUser() user: { id: string },
    @Body()
    createStatusDto: {
      type: "TEXT" | "IMAGE" | "VIDEO";
      content?: string;
      mediaUrl?: string;
      duration?: number;
    },
  ) {
    return this.statusService.createStatus(user.id, createStatusDto);
  }

  @Get(":id")
  @ApiOperation({ summary: "Get status" })
  async getStatus(@Param("id") id: string) {
    return this.statusService.getStatusById(id);
  }

  @Delete(":id")
  @HttpCode(HttpStatus.NO_CONTENT)
  @ApiOperation({ summary: "Delete status" })
  async deleteStatus(
    @CurrentUser() user: { id: string },
    @Param("id") id: string,
  ) {
    await this.statusService.deleteStatus(id, user.id);
  }

  @Post(":id\\:view")
  @ApiOperation({ summary: "Mark status viewed" })
  async viewStatus(
    @CurrentUser() user: { id: string },
    @Param("id") id: string,
  ) {
    return this.statusService.viewStatus(id, user.id);
  }
}

File: src/main/java/com/chat/status/controller/StatusController.java
package com.chat.status.controller;

import com.chat.status.service.StatusService;
import java.util.Map;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/v1/status")
public class StatusController {

  private final StatusService statusService;

  public StatusController(StatusService statusService) {
    this.statusService = statusService;
  }

  @GetMapping
  public Map<String, Object> list() {
    return statusService.list();
  }

  @PostMapping
  @ResponseStatus(HttpStatus.CREATED)
  public Map<String, Object> create(
      Authentication authentication, @RequestBody Map<String, Object> body) {
    return statusService.create(
        authentication.getName(),
        body.get("content") == null ? null : String.valueOf(body.get("content")),
        body.get("mediaUrl") == null ? null : String.valueOf(body.get("mediaUrl")),
        body.get("statusType") == null ? "TEXT" : String.valueOf(body.get("statusType")));
  }

  @GetMapping("{status}")
  public Map<String, Object> get(@PathVariable String status) {
    return statusService.get(status);
  }

  @DeleteMapping("{status}")
  @ResponseStatus(HttpStatus.NO_CONTENT)
  public void delete(Authentication authentication, @PathVariable String status) {
    statusService.delete(status, authentication.getName());
  }

  @PostMapping("{status}:view")
  @ResponseStatus(HttpStatus.NO_CONTENT)
  public void view(Authentication authentication, @PathVariable String status) {
    statusService.view(status, authentication.getName());
  }
}
